### PR TITLE
Update getPullRequestsMergedBetween logic

### PR DESCRIPTION
**Pull request by roryabraham**
cc @mountiny @AndrewGable @Jag96 

### Details
This (hopefully) fixes long-broken logic where duplicate PRs show up in checklists / deploy comments or are excluded from checklists / deploy comments in error.

See [this comment](https://github.com/Expensify/App/issues/4977#issuecomment-1001656486) for details.

### Fixed Issues
$ https://github.com/Expensify/App/issues/4977

### Tests
I wrote an automated test script to run these steps on an ongoing basis, but here are the manual testing steps:

#### Setup steps

1. If needed, install `jq`: `brew install jq`.
1. Create directory for dummy repo: `mkdir DummyRepo && cd DummyRepo`
1. Initialize npm project so that we can test the node script w/ underscore dependency: `npm init` (use `0.0.1` as the initial version).
1. Install underscore: `npm install underscore && npm i`
1. Create a tiny node script we'll use for testing: `touch ./getPullRequestsMergedBetween.js && chmod u+x ./getPullRequestsMergedBetween.js`

    ```javascript
    #!/usr/bin/env node

    # Note: you'll need to use your local path to the Expensify/App repo
    const GitUtils = require('../Expensidev/App/.github/libs/GitUtils.js');

    GitUtils.getPullRequestsMergedBetween(process.argv[2], process.argv[3]);
    ```

1. Initialize `DummyRepo` as a git repo with `main` as the default branch.

    ```bash
    git init -b main
    git add package.json package-lock.json getPullRequestsMergedBetween.js
    git commit -m "Initial commit"
    ```
    
1. Update the version to `1.0.0` and tag the commit. This is only necessary so that we have a "previous tag".

    ```bash
    npm --no-git-tag-version version 1.0.0 -m "Update version to 1.0.0"
    git add package.json
    git commit -m "Update version to 1.0.0"
    ```

1. Create `staging` and `production` branches off of `main`:

    ```bash
    git checkout -b staging
    git checkout -b production
    git checkout main
    ```
    
1. Tag the version on staging:

    ```bash
    git checkout staging
    git tag $(cat ./package.json | jq -r .version)
    git checkout main
    ```

#### Scenario 1: Merge a pull request when the checklist is unlocked

1. Create "PR 1", and merge that PR to `main`.

    ```bash
    git checkout main
    git checkout -b pr-1
    echo "Changes from PR #1" >> CHANGELOG.txt 
    git add CHANGELOG.txt
    git commit -m "Create changelog"
    git checkout main
    git merge pr-1 --no-ff -m "Merge pull request #1 from Expensify/pr-1"
    git branch -d pr-1
    ```
    
1. Bump the version to `1.0.1`:

    ```bash
    git checkout main
    git checkout -b version-bump
    npm --no-git-tag-version version 1.0.1 -m "Update version to 1.0.1"
    git add package.json package-lock.json 
    git commit -m "Update version to $(cat package.json | jq -r .version)"
    git checkout main
    git merge version-bump --no-ff -m "Merge pull request #2 from Expensify/version-bump"
    git branch -d version-bump
    ```
    
1. Merge `main` into `staging`:

    ```bash
    git checkout staging
    git checkout -b update-staging-from-main
    git merge -Xtheirs main
    git checkout staging
    git merge update-staging-from-main --no-ff -m "Merge pull request #3 from Expensify/update-staging-from-main"
    git branch -d update-staging-from-main
    ```
    
1. Tag `staging`: 

    ```bash
    git checkout staging
    git tag $(cat ./package.json | jq -r .version)
    git checkout main
    ```

1. From `staging`, verify that `node getPullRequestsMergedBetween.js 1.0.0 1.0.1` outputs:

	  ```
	  List of pull requests merged between 1.0.0 and 1.0.1 [ '1', ]
	  ```

#### Scenario 2: Lock the deploy checklist

1. Bump the version to `1.1.0` on `main`

    ```bash
    git checkout main
    git checkout -b version-bump
    npm --no-git-tag-version version 1.1.0 -m "Update version to 1.1.0"
    git add package.json package-lock.json
    git commit -m "Update version to $(cat package.json | jq -r .version)"
    git checkout main
    git merge version-bump --no-ff -m "Merge pull request #4 from Expensify/version-bump"
    git branch -d version-bump
    ```
    
1. Merge `main` into `staging`

    ```bash
    git checkout staging
    git checkout -b update-staging-from-main
    git merge -Xtheirs main
    git checkout staging
    git merge update-staging-from-main --no-ff -m "Merge pull request #5 from Expensify/update-staging-from-main"
    git branch -d update-staging-from-main
    ```

1. Tag the new version on `staging`

    ```bash
    git checkout staging
    git tag $(cat package.json | jq -r .version)
    ```

1. From `staging`, verify that `node getPullRequestsMergedBetween 1.0.0 1.1.0` outputs:

    ```
    List of pull requests merged between 1.0.0 and 1.1.0 [ '1' ]
    ```

#### Scenario 3: Merge another pull request, but don't CP it

1. Create "PR 6", and merge that PR to `main`:

    ```bash
    git checkout main
    git checkout -b pr-6
    echo "Changes from PR #6" >> CHANGELOG.txt 
    git add CHANGELOG.txt
    git commit -m "Update changelog for PR #6"
    git checkout main
    git merge pr-6 --no-ff -m "Merge pull request #6 from Expensify/pr-6"
    git branch -d pr-6
    ```
    
#### Scenario 4: Merge another pull request and CP it to staging

1. Create "PR 7" and merge that PR to `main`:

    ```bash
    git checkout main
    git checkout -b pr-7
    echo "\nChanges from PR #7\n" >> CHANGELOG.txt 
    git add CHANGELOG.txt
    git commit -m "Update changelog for PR #7"
    git checkout main
    git merge pr-7 --no-ff -m "Merge pull request #7 from Expensify/pr-7"
    git branch -d pr-7
    ```
    
1. Update version to `1.1.1` on `main`:

    ```bash
    git checkout main
    git checkout -b version-bump
    npm --no-git-tag-version version 1.1.1 -m "Update version to 1.1.1"
    git add package.json package-lock.json
    git commit -m "Update version to $(cat package.json | jq -r .version)"
    git checkout main
    git merge version-bump --no-ff -m "Merge pull request #8 from Expensify/version-bump"
    git branch -d version-bump
    ```
    
1. Cherry pick PR 7 (and the version-bump) to `staging`:

    1. First, find the merge commit hash for PR 7 using `git log`. Save it to a local variable called `MERGE_7`
    1. Next, find the merge commit hash for PR 8 using `git log`. Save it to a local variable called `MERGE_8`
    1. Next, begin using the `cherry-pick` command to CP our changes:
    
        ```bash
	     git checkout staging
	     git checkout -b cherry-pick-staging-7
	     git cherry-pick -S -x --mainline 1 --strategy=recursive -Xtheirs $MERGE_8
	     git cherry-pick -S -x --mainline 1 $MERGE_7
	     ```
	 
	 1. At this point, there will be conflicts in `CHANGELOG.txt`. Fix those conflicts to include only the ones to CP (i.e.: "changes from PR 7").
	 
	     ```bash
	     # manually resolve conflicts
	      vim CHANGELOG.txt
	      git add CHANGELOG.txt
	      
	      # Then continue the CP
	      git cherry-pick --continue
	      
	      # And merge the pretend CP-PR
	      git checkout staging
	      git merge cherry-pick-staging-7 --no-ff -m "Merge pull request #9 from Expensify/cherry-pick-staging-7"
	      git branch -d cherry-pick-staging-7
          ```

1. Tag the new version on `staging`:

    ```bash
    git checkout staging
    git tag $(cat package.json | jq -r .version)
    ```

1. From `staging`, verify that `node getPullRequestsMergedBetween 1.0.0 1.1.1` outputs:

    ```
    List of pull requests merged between 1.0.0 and 1.1.1 [ '9', '7', '1' ]
    ```
    
    **Note:** It's fine that PR 9 is included here. It will be filtered out of the deploy checklist at a later step because it's created by `OSBotify`.

##### Scenario 5A: Run the production deploy

1. Update production from staging

    ```bash
    git checkout production
    git checkout -b update-production-from-staging
    git merge -Xtheirs staging
    git checkout production
    git merge update-production-from-staging --no-ff -m "Merge pull request #10 from Expensify/update-production-from-staging"
    git branch -d update-production-from-staging
    ```
    
1. Verify that `node getPullRequestsMergedBetween 1.0.0 1.1.1` outputs:

	  ```
	  List of pull requests merged between 1.0.0 and 1.1.1 [ '9', '7', '1' ]
	  ```
		   
##### Scenario 5B: Run the staging deploy and create a new checklist

1. Bump version to `1.1.2` on `main`:

    ```bash
    git checkout main
    git checkout -b version-bump
    npm --no-git-tag-version version 1.1.2 -m "Update version to 1.1.2"
    git add package.json package-lock.json
    git commit -m "Update version to $(cat package.json | jq -r .version)"
    git checkout main
    git merge version-bump --no-ff -m "Merge pull request #11 from Expensify/version-bump"
    git branch -d version-bump
    ```

1. Update `staging` from `main`:

    ```bash
    git checkout staging
    git checkout -b update-staging-from-main
    git merge -Xtheirs main
    git checkout staging
    git merge update-staging-from-main --no-ff -m "Merge pull request #12 from Expensify/update-staging-from-main"
    git branch -d update-staging-from-main
    ```

1. Tag new version on `staging`:

    ```bash
    git checkout staging
    git tag $(cat package.json | jq -r .version)
    ```
    
1. Verify that `node getPullRequestsMergedBetween 1.1.1 1.1.2` outputs:

    ```
    List of pull requests merged between 1.1.1 and 1.1.2 [ '9', '6' ]
    ```

#### Scenario 6: Merge another pull request with the checklist unlocked

1. Create "PR 13", and merge that PR to `main`.

    ```bash
    git checkout main
    git checkout -b pr-13
    echo "Changes from PR #13" >> CHANGELOG.txt 
    git add CHANGELOG.txt
    git commit -m "Update changelog in PR 13"
    git checkout main
    git merge pr-13 --no-ff -m "Merge pull request #13 from Expensify/pr-13"
    git branch -d pr-13
    ```
    
1. Bump the version to `1.1.3`:

    ```bash
    git checkout main
    git checkout -b version-bump
    npm --no-git-tag-version version 1.1.3 -m "Update version to 1.1.3"
    git add package.json package-lock.json 
    git commit -m "Update version to $(cat package.json | jq -r .version)"
    git checkout main
    git merge version-bump --no-ff -m "Merge pull request #14 from Expensify/version-bump"
    git branch -d version-bump
    ```
    
1. Merge `main` into `staging`:

    ```bash
    git checkout staging
    git checkout -b update-staging-from-main
    git merge -Xtheirs main
    git checkout staging
    git merge update-staging-from-main --no-ff -m "Merge pull request #15 from Expensify/update-staging-from-main"
    git branch -d update-staging-from-main
    ```
    
1. Tag `staging`: 

    ```bash
    git checkout staging
    git tag $(cat ./package.json | jq -r .version)
    ```

1. From `staging`, verify that `node getPullRequestsMergedBetween.js 1.1.1 1.1.3` outputs:

	  ```
	  List of pull requests merged between 1.1.1 and 1.1.3 [ '13', '9', '6' ]
	  ```
	  
---

In addition to the above, I updated some automated unit tests.

### QA Steps
Here are the live-testing steps for this PR... We'll need 5 PRs A-E.

1. With the deploy checklist unlocked, merge PR A. Then:
    - [ ] PR A should be deployed to staging
    - [ ] PR A should be added to the deploy checklist
    - [ ] PR A should receive a "deployed to staging" comment
1. With the deploy checklist unlocked, merge PR B. Then:
    - [ ] PR B should be deployed to staging
    - [ ] PR B should be added to the deploy checklist
    - [ ] PR B should receive a "deployed to staging" comment
    - [ ] PR A _should not_ receive another "deployed to staging" comment.
1. Lock the checklist.
    - [ ] PR A _should not_ receive another "deployed to staging" comment
    - [ ] PR B _should not_ receive another "deployed to staging" comment
1. Merge PR C _without_ `CP Staging` label
    - [ ] Deploy should not happen
    - [ ] PR C should receive "not yet deployed" comment
1. Merge PR D _with_ `CP Staging` label
    - [ ] PR D should be deployed to staging
    - [ ] PR D should be added to the deploy checklist
    - [ ] Applauseleads should be tagged to be informed of the newly-CP'd PR.
    - [ ] PR D should receive a "CP'd to staging" comment
1. Merge PR E _without_ `CP Staging` label, then manually CP it.
    - [ ] PR E should be deployed to staging
    - [ ] PR E should be added to the deploy checklist
    - [ ] Applauseleads should be tagged to be informed of the newly-CP'd PR.
    - [ ] PR E should receive a "CP'd to staging" comment
1. Close the checklist
    - [ ] Production deploy should happen
    - [ ] PRs `A`, `B`, `D`, `E` should all receive "deployed to production" comment
    - [ ] PR `C` should _not_ receive "deployed to production" comment.
    - [ ] Staging deploy should also happen
    - [ ] PR `C` should be included in the new checklist
    - [ ] PRs `A`, `B`, `D`, `E` should all _not_ be included in the new checklist.
    - [ ] PR `C` should receive "deployed to staging" comment.
    - [ ] PRs `A`, `B`, `D`, `E` should all _not_ receive "deployed to staging" comment.

### Tested On
GitHub only

### CODE REVIEW

**Comment by roryabraham** [comment]
Question: Is it valuable to aggregate those many long testing steps into an automated script? Should be pretty easy... The only tricky part is the CP conflicts, but those aren't super critical and I can change the steps a bit so there's no conflict.

**Comment by roryabraham** [comment]
Sorry, marking this as WIP. I created an automated testing script and it's passing locally, but on the GitHub Actions runner I'm getting this error when attempting the `git cherry-pick` command:

```
error: gpg failed to sign the data
error: failed to write commit object
fatal: cherry-pick failed
```

I'm not sure why it's requiring us to GPG-sign the data. That really shouldn't be necessary because it's in a separate dummy repo, not in E/App.

**Comment by roryabraham** [comment]
Yay it's working 🎉

**Comment by roryabraham** [comment]
Okay, this is definitely ready for review now

**Review by AndrewGable — DISMISSED** [review]
This looks great!! 😍

**Review by Jag96 — DISMISSED** [review]
Looks great! Just a couple of nitpicks, excited for these changes!

File: shellUtils.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  echo "💥 $RED$1$RESET"
+}
+
+function info {
+  echo "$BLUE$1$RESET"
+}
+
+function title {
+  echo "
+$TITLE$1$RESET"
```
[the remark this answers is not in the document]
Comment (roryabraham) [reply shellUtils.sh:23]: This is intentional – I would use a `\n` but echo doesn't reliably expand escape sequences. I can use `printf` instead, if you prefer. That would look like:

```bash
function title {
  printf "\n%s%s%s\n" "$TITLE" "$1" "$RESET"
}
```

[the remark this answers is not in the document]
Comment (roryabraham) [reply shellUtils.sh:23]: I personally feel like the newline literal is a bit simpler/more obvious, but I could go either way with it.

[the remark this answers is not in the document]
Comment (roryabraham) [reply shellUtils.sh]: Updated to use `printf` instead.

File: .github/actions/createOrUpdateStagingDeploy/index.js [hunk 2]
```diff
@@ -188,43 +188,80 @@ const _ = __nccwpck_require__(3571);
...
-    // Additionally, we omit merge commits made while cherry picking using negative lookahead in the regexp.
-    const pullRequestIDs = _.reduce(commitMessages, (mergedPRs, commitMessage) => {
-        const mergeCommits = [
-            ...commitMessage.matchAll(/Merge pull request #(\d{1,6}) from (?!(?:Expensify\/(?:master|main|version-))|(?:([\s\S]*?)\(cherry picked from commit .*\)\s*))/gm),
-        ];
+/**
+ * Parse merged PRs, excluding those from irrelevant branches.
+ *
+ * @param {Array<String>} commitMessages
+ * @returns {*}
```
[the remark this answers is not in the document]
Comment (roryabraham) [reply .github/actions/createOrUpdateStagingDeploy/index.js]: Good catch!

**Review by mountiny — APPROVED** [review]
Wohoo 🎉  This is absolutely great work and investigation here. I haven't found any typos or any changes I would suggest. I believed you have tested this thoroughly so going to approve it now. 

There is one thing I can think of and that is some more extensive description of this entire process and why we need to do it (parse and filter the commits in this way). Since it took us a long time to figure it out and hopefully, this will work and solve our issues, would you write some SO to cover this for future engineers to understand better?

Thank you for working on this!

**Review by Jag96 — APPROVED** [review]

**Review by ctkochan22 — APPROVED** [review]
NAB, we repeat some code in all of these instances that could probably be dry'd up, but it was like that before.

**Comment by botify** [comment]
@ctkochan22 looks like this was merged without passing tests. Please add a note explaining why this was done and remove the `Emergency` label if this is not an emergency.

**Comment by ctkochan22** [comment]
Double checked with @Jag96 , this sounds like https://github.com/Expensify/Expensify/issues/189984

Removing emegency label

**Comment by roryabraham** [comment]
Okay, I've gathered together some relatively low-risk PRs we can use to QA this:

**A:** https://github.com/Expensify/App/pull/7072
**B:** https://github.com/Expensify/App/pull/7137
**C:** https://github.com/Expensify/App/pull/7154
**D:** https://github.com/Expensify/App/pull/7149

**Comment by roryabraham** [comment]
Updated the QA steps to account for https://github.com/Expensify/App/pull/7161 – so the new 5 PRs will be:

**A**: #7161
**B**: #7072
**C**: #7137 
**D**: #7154 
**E**: #7149

**Comment by roryabraham** [comment]
Okay, starting QA using the 5 pull requests as listed in the previous comment:

1. With the deploy checklist unlocked, merge PR A. Then:
    - [x] PR A should be deployed to staging
    - [x] PR A should be added to the deploy checklist
    - [ ] PR A should receive a "deployed to staging" comment
1. With the deploy checklist unlocked, merge PR B. Then:
    - [ ] PR B should be deployed to staging
    - [ ] PR B should be added to the deploy checklist
    - [ ] PR B should receive a "deployed to staging" comment
    - [ ] PR A _should not_ receive another "deployed to staging" comment.
1. Lock the checklist.
    - [ ] PR A _should not_ receive another "deployed to staging" comment
    - [ ] PR B _should not_ receive another "deployed to staging" comment
1. Merge PR C _without_ `CP Staging` label
    - [ ] Deploy should not happen
    - [ ] PR C should receive "not yet deployed" comment
1. Merge PR D _with_ `CP Staging` label
    - [ ] PR D should be deployed to staging
    - [ ] PR D should be added to the deploy checklist
    - [ ] Applauseleads should be tagged to be informed of the newly-CP'd PR.
    - [ ] PR D should receive a "CP'd to staging" comment
1. Merge PR E _without_ `CP Staging` label, then manually CP it.
    - [ ] PR E should be deployed to staging
    - [ ] PR E should be added to the deploy checklist
    - [ ] Applauseleads should be tagged to be informed of the newly-CP'd PR.
    - [ ] PR E should receive a "CP'd to staging" comment
1. Close the checklist
    - [ ] Production deploy should happen
    - [ ] PRs `A`, `B`, `D`, `E` should all receive "deployed to production" comment
    - [ ] PR `C` should _not_ receive "deployed to production" comment.
    - [ ] Staging deploy should also happen
    - [ ] PR `C` should be included in the new checklist
    - [ ] PRs `A`, `B`, `D`, `E` should all _not_ be included in the new checklist.
    - [ ] PR `C` should receive "deployed to staging" comment.
    - [ ] PRs `A`, `B`, `D`, `E` should all _not_ receive "deployed to staging" comment.

**Comment by roryabraham** [comment]
Oh, and `Applauseleads should be tagged to be informed of the newly CP'd PR` was aspirational – I wrote the QA steps before completing the pull request and never implemented this requirement. We can handle it in https://github.com/Expensify/App/issues/7178

**Comment by roryabraham** [comment]
Finished QA here 🎉

**Comment by mountiny** [comment]
@roryabraham Amazing job on this one Rory!

**Pull request merged**